odoo/tools/sql.py

```diff
@@ -405,8 +405,6 @@ def constraint_definition(cr, tablename, constraintname):
 
 def add_constraint(cr, tablename, constraintname, definition):
     """ Add a constraint on the given table. """
-    if "%" in definition:
-        definition = definition.replace("%", "%%")
```

Comment:
These lines were here to avoid misinterpretation of potential `%s` seen as a positional argument in the `SQL.__init__`. We don't check it in **17.0** if there isn't any `args` given to SQL, but we do in later version (added here https://github.com/odoo/odoo/pull/173794 in saas17.5: `code % ()`).

Then for the 18.0+ versions: I think you need to keep this but moving it at the creation of query1 (`SQL(definition)` => `SQL(definition.replace('%', '%%')`) but also modify the code of `SQL.__init__` (`code % ()` => `code = code % ()`).

Or I miss something ?

Reply:
In 18 the patch needs to be adapted.

Reply:
> These lines were here to avoid misinterpretation of potential %s seen as a positional argument in the SQL.__init__. We don't check it in 17.0 if there isn't any args given to SQL, but we do in later version (added here https://github.com/odoo/odoo/pull/173794 in saas17.5: code % ()).

For constraint definitions I think we do not support any sort of parameters, thus I believe this should be fine.

> Then for the 18.0+ versions: I think you need to keep this but moving it at the creation of query1 (SQL(definition) => SQL(definition.replace('%', '%%')) but also modify the code of SQL.__init__ (code % () => code = code % ()).
>
> Or I miss something ?

I tested your patch proposal and it works in 18. The current test also fails in 18 with the patch here. 
```
2024-11-26 13:26:30,985 294224 ERROR test_18 odoo.addons.base.tests.test_sql: FAIL: TestSqlTools.test_add_constraint
Traceback (most recent call last):
  File "/home/odoo/src/odoo/18.0/odoo/addons/base/tests/test_sql.py", line 171, in test_add_constraint
    with self.assertRaises(CheckViolation), mute_logger('odoo.sql_db'):
  File "/usr/lib/python3.12/contextlib.py", line 144, in __exit__
    next(self.gen)
  File "/home/odoo/src/odoo/18.0/odoo/tests/common.py", line 473, in _assertRaises
    with ExitStack() as inner:
  File "/usr/lib/python3.12/contextlib.py", line 610, in __exit__
    raise exc_details[1]
  File "/usr/lib/python3.12/contextlib.py", line 595, in __exit__
    if cb(*exc_details):
       ^^^^^^^^^^^^^^^^
AssertionError: CheckViolation not raised
```

I think we can make the update in the forward port. wdyt?

Reply:
> I think we can make the update in the forward port. wdyt?

Sound good. Thanks for fix.